src/core/border_router/routing_manager.cpp

```diff
@@ -1042,69 +1104,82 @@ bool RoutingManager::UpdateDiscoveredPrefixes(const RouterAdv::RouteInfoOption &
 {
-    uint8_t         removedNum          = 0;
-    ExternalPrefix *keptPrefix          = mDiscoveredPrefixes;
-    TimeMilli       now                 = TimerMilli::GetNow();
-    TimeMilli       earliestExpireTime  = now.GetDistantFuture();
-    uint8_t         keptOnLinkPrefixNum = 0;
+    uint8_t         removedNum               = 0;
+    ExternalPrefix *remainingPrefix          = mDiscoveredPrefixes;
+    TimeMilli       now                      = TimerMilli::GetNow();
+    uint8_t         remainingOnLinkPrefixNum = 0;
+
+    mDiscoveredPrefixInvalidTimer.Stop();
 
     for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
         ExternalPrefix &prefix = mDiscoveredPrefixes[i];
 
         if ((aPrefix != nullptr && prefix.mPrefix == *aPrefix && prefix.mIsOnLinkPrefix == aIsOnLinkPrefix) ||
-            (prefix.mExpireTime <= now))
+            (prefix.GetExpireTime() <= now))
         {
             RemoveExternalRoute(prefix.mPrefix);
             ++removedNum;
         }
         else
         {
-            earliestExpireTime = OT_MIN(earliestExpireTime, prefix.mExpireTime);
-            *keptPrefix        = prefix;
-            ++keptPrefix;
+            mDiscoveredPrefixInvalidTimer.FireAtIfEarlier(prefix.GetExpireTime());
+            *remainingPrefix = prefix;
+            ++remainingPrefix;
             if (prefix.mIsOnLinkPrefix)
             {
-                ++keptOnLinkPrefixNum;
+                ++remainingOnLinkPrefixNum;
             }
         }
     }
 
     mDiscoveredPrefixNum -= removedNum;
 
-    if (keptOnLinkPrefixNum == 0)
-    {
-        mDiscoveredPrefixInvalidTimer.Stop();
-
-        if (mAdvertisedOnLinkPrefix == nullptr)
-        {
-            // There are no valid on-link prefixes on infra link now, start Router Solicitation
-            // To find out more on-link prefixes or timeout to advertise my local on-link prefix.
-            StartRouterSolicitationDelay();
-        }
-    }
-    else
+    if (remainingOnLinkPrefixNum == 0 && mAdvertisedOnLinkPrefix == nullptr)
     {
-        mDiscoveredPrefixInvalidTimer.FireAt(earliestExpireTime);
+        // There are no valid on-link prefixes on infra link now, start Router Solicitation
+        // To discover more on-link prefixes or timeout to advertise my local on-link prefix.
+        StartRouterSolicitationDelay();
     }
 
     return (removedNum != 0); // If anything was removed we need to reevaluate.
 }
 
 void RoutingManager::InvalidateAllDiscoveredPrefixes(void)
 {
-    TimeMilli past = TimerMilli::GetNow();
-
     for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
-        mDiscoveredPrefixes[i].mExpireTime = past;
+        mDiscoveredPrefixes[i].mValidLifetime = 0;
     }
 
     InvalidateDiscoveredPrefixes();
 
     OT_ASSERT(mDiscoveredPrefixNum == 0);
 }
 
+static void UpdateStaleTime(TimeMilli &aStaleTime,
+                            bool       aStaleTimeInitialized,
```

Comment:
Is this supposed to be pass-by-reference?

Reply:
You are right, thanks!